### ISSUE TITLE
Configuration of SST.

### ISSUE THREAD

**Issue by bradysprings** [introlab/odas#78]
Hello @FrancoisGrondin ,

I have changed the sampling rate (48000hz), hop size (2048), frame size (4096) to do some experiment. SSL works well but SST no longer works (the big highlighted dots do not appear any more). 

I think it may be because of the number of pots being generated (from SSL) per unit time decreased. 

-     Previous setting - Each frame : 4 pots every 16ms (original setting: 16000hz, 128, 256)
-     Current setting - Each frame : 4 pots every 85ms (new setting: 48000hz, 2048, 4096)

Or could there be any other reason? Could you help me find the correct configuration to make SST work with my current settings?

Thanks!

**Comment by FrancoisGrondin** [comment]
Yep could be that. Have you play with the detection threshold?

**Comment by bradysprings** [comment]
How do I play with the detection threshold?
Could you let me know which configuration (https://github.com/introlab/odas/wiki/configuration) variable I should play with?

**Comment by FrancoisGrondin** [comment]
Sorry for the late reply:

The energy distribution of the localization process is modeled by two GMMs, one for source activity, and one for source inactivity. In most cases you can simply have one gaussian for each, and their parameters are given in `active.[0].mu` and `inactive.[0].mu`. You can usually decrease the mean of active to make detection more sensitive.

**Comment by bradysprings** [comment]
Reducing active.[0].mu and inactive.[0].mu did not work well. I also tried increasing the number of pots in the ssl section of the configuration file from 4 to 8 but got a segmentation fault. Do I need to change something else together?

**Comment by FrancoisGrondin** [comment]
Also, please try to reduce the frameSize to 1024, and set the hopSize to 256. Tell me if it helps.

**Comment by luc2yj** [comment]
Hello, @FrancoisGrondin  

Currently, changing the number of pots causes a segmentation fault. Do you have plans to fix the issue?

Thanks!

**Comment by FrancoisGrondin** [comment]
Ok; I'm currently on vacations: let me address that next week. Out of curiosity, when you get the segfault, what's the number of potential sources?

**Comment by luc2yj** [comment]
Thank you @FrancoisGrondin ! Anything other than 4 causes a segmentation fault.

**Comment by FrancoisGrondin** [comment]
:O I can't believe I missed that. Will definitely investigate. Thanks for pointing this out!

**Comment by bradysprings** [comment]
Could you leave a comment in this thread once you fix it? Thanks.

**Comment by luc2yj** [comment]
Sounds good @FrancoisGrondin !
I tracked the problem just a bit and it seems like it runs well until ODAS runs the amod_sst_thread function in amod_sst.c. More specifically, within the function the line below seems to generate the segmentation fault.
            msg_pots_in = amsg_pots_filled_pop(obj->in1);
I didn't go further from here though.
Please let me know how you fixed it as I want to modify my code as well.

**Comment by FrancoisGrondin** [comment]
Thank you! I'll run valgrind and investigate 'til I find the source of the problem!

**Comment by bradysprings** [comment]
When do you think you will have this issue fixed? Is there an estimate of fixing this error? Thank you @FrancoisGrondin .

**Comment by FrancoisGrondin** [comment]
So I ran odas with valgrind for the pheonix branch (which contains the version 3.0 of odas, with many new features, and I don't get any segfault while setting nPots to a different value). I suppose you are experiencing this problem with the version on the master branch? Did any of view tried with version 3.0 in the phoenix branch?

**Comment by luc2yj** [comment]
Yes. I am using the master branch. I will check out the phoenix branch. Hopeful the code is similar so that I can  easily port the modifications that I made.

**Comment by luc2yj** [comment]
@FrancoisGrondin , it seems like the phoenix branch is quite different from the master branch. To begin with, it looks like the matrix voice configuration file in the phoenix branch is empty. 

How do I port the current configuration file to the phoenix branch? Would directly using the current configuration file work? 

Also, does it still work with ODAS web?

How do I set the port for reading the raw file. Is it set to 5001 by default in most cases? It seems like the master branch version didn't require such information.

Can we use the phoenix version like the odaslive? instead of the odasserver?

**Comment by bradysprings** [comment]
I have been using the odas master branch as well and was using odaslive. Is it possible to fix the error in the master branch as well?

**Comment by FrancoisGrondin** [comment]
Yes, the phoenix branch is a brand new thing with odas. It is very different from master, and this is why I'm not merging it yet. My goal is to have a first stable release based on the pheonix branch. Here is to have odas behaves as a server that accepts socket connections, and uses external tools to format the data. 

I plan to adapt every config file to match the phoenix config files (which are now in JSON). I wrote my own JSON parser, so this way I removed dependency on libconfig. Moreover, using sockets removes the needs to use ALSA (you can for instance push the raw with arecord, or any other software, which makes the library OS independent).

My colleague will have to update odas web to make it compatible with phoenix.

For now I'm aware many people use the current version in master, so I will try to reproduce your bug in the master branch tomorrow, and fix it.

**Comment by bradysprings** [comment]
Your latest master branch version seems to have no problems. Let me port my code to you latest master branch and see how it works. @luc2yj you should update it to the latest master branch as well.

**Comment by FrancoisGrondin** [comment]
@bradysprings Oh! Lemme know if the problem is solved with the latest version

**Comment by luc2yj** [comment]
@bradysprings Thank you for letting me know. It did help. 👍 

@FrancoisGrondin  It does work until nPots=9. Starting from nPots=10 a segmentation fault occurs. 

I found this problem by testing the configuration fields of SSL because I wanted to exactly understand how the potential points were getting plotted on ODAS_WEB. Since you paper is accepted, can you provide more details regarding what the configurations mean? By reading ODAS code, I did understand what some of them do. But if you could explain the effect of the parameters it would be very helpful.

nPots: the number of potential points to get from GCC PHAT result, 
nMatches: Not sure, 
probMin: Not sure, 
nRefinedLevels: Not sure, 
interpRate: used in ssl to interpolate the freq information before doing ifft,	    
scans.[0].level,	
scans.[0].delta,
scans.[1].level,
scans.[2].level

**Comment by FrancoisGrondin** [comment]
Yes, I'm now doing the minor revision on the paper, and I'll push it on Arxiv in the next few days.
